stan/math/rev/fun/softmax.hpp

```diff
@@ -24,24 +24,46 @@ namespace math {
  * @return Softmax of the input.
  * @throw std::domain_error If the input vector is size 0.
  */
-
 inline Eigen::Matrix<var, Eigen::Dynamic, 1> softmax(
     const Eigen::Matrix<var, Eigen::Dynamic, 1>& alpha) {
-  check_nonzero_size("softmax", "alpha", alpha);
+  if (alpha.size() == 0) {
+    return alpha;
+  }
 
   arena_matrix<Eigen::VectorXd> res_val = softmax(value_of(alpha));
   arena_matrix<Eigen::Matrix<var, Eigen::Dynamic, 1>> res = res_val;
   arena_matrix<Eigen::Matrix<var, Eigen::Dynamic, 1>> alpha_arena = alpha;
 
-  reverse_pass_callback([=]() mutable {
+  reverse_pass_callback([res_val, res, alpha_arena]() mutable {
     const auto& res_adj = to_ref(res.adj());
     alpha_arena.adj()
-        = -res_val * res_adj.dot(res_val) + res_val.cwiseProduct(res_adj);
+        += -res_val * res_adj.dot(res_val) + res_val.cwiseProduct(res_adj);
   });
 
   return res;
 }
 
+/**
+ * Return the softmax of the specified `var_value<vector>`. Softmax is
+ * guaranteed to return a simplex.
+ *
+ * @param alpha Unconstrained input vector.
+ * @return Softmax of the input.
+ * @throw std::domain_error If the input vector is size 0.
+ */
+template <typename Mat, require_var_vt<is_eigen, Mat>* = nullptr>
+inline auto softmax(const Mat& alpha) {
+  if (alpha.size() == 0) {
+    return alpha;
+  }
+  Mat res(softmax(value_of(alpha)));
```

Comment:
Looks good. I think this should use the same code as the other version though.

This is the other version:
```cpp
arena_matrix<Eigen::VectorXd> res_val = softmax(value_of(alpha));
arena_matrix<Eigen::Matrix<var, Eigen::Dynamic, 1>> res = res_val;
arena_matrix<Eigen::Matrix<var, Eigen::Dynamic, 1>> alpha_arena = alpha;
```

If the input argument is type `const T&`, we can do something like:
```cpp
const arena_matrix<Eigen::VectorXd> res_val = softmax(value_of(alpha));
arena_t<T> res = res_val;
arena_t<T> alpha_arena = alpha;
```

And for the `var_value<Eigen>` version `arena_t` just returns `var_value` types. The extra trick to avoid copying is that:

```cpp
var_value<Eigen::VectorXd> res = res_val;
```

should recognize that `res_val` is a const arena_matrix and so it's safe to just take the value pointer in. We're using that assumption in this constructor: https://github.com/stan-dev/math/blob/develop/stan/math/rev/functor/arena_matrix.hpp#L71 for instance

Reply:
> And for the var_value<Eigen> version arena_t just returns var_value types. The extra trick to avoid copying is that:

Just to make sure we are on the same page, are you saying that we should have an `arena_matrix<Eigen::Matrix<double>>` constructor for `var_value` and `vari_value` that just takes in the `arena_matrix<Eigen::Matrix<double>>` and assigns the memory it's pointing to the `vari_value`'s `val_mem`? I think that makes sense to me

Reply:
I think it would be a `const arena_matrix<Eigen::Matrix<double>>&` constructor, but yeah, it can use the same memory since it's constant.